src/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -88,37 +85,66 @@ public static class HostWriter
+                0x72, 0x7b, 0x93, 0x02, 0x14, 0xd7, 0xa0, 0x32,
+                0x13, 0xf5, 0xb9, 0xe6, 0xef, 0xae, 0x33, 0x18,
+                0xee, 0x3b, 0x2d, 0xce, 0x24, 0xb3, 0x6a, 0xae
+            };
 
             // Re-write the destination apphost with the proper contents.
-            BinaryUtils.SearchAndReplace(appHostDestinationFilePath, BundleHeaderPlaceholderSearchValue, BitConverter.GetBytes(bundleHeaderOffset));
+            BinaryUtils.SearchAndReplace(appHostPath,
+                                         bundleHeaderPlaceholder,
+                                         BitConverter.GetBytes(bundleHeaderOffset), 
+                                         pad0s:false);
 
             // Memory-mapped write does not updating last write time
-            File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            File.SetLastWriteTimeUtc(appHostPath, DateTime.UtcNow);
         }
 
-        private static void CopyAppHost(
-            string appHostSourceFilePath,
-            string appHostDestinationFilePath)
+        /// <summary>
+        /// Check if the an AppHost is a single-file bundle
+        /// </summary>
+        /// <param name="appHostFilePath">The path of Apphost to check</param>
+        /// <param name="bundleHeaderOffset">An out parameter containing the offset of the bundle header (if any)</param>
+        /// <returns>True if the AppHost is a single-file bundle, false otherwise</returns>
+        public static bool IsBundle(string appHostFilePath, out long bundleHeaderOffset)
         {
-            var destinationDirectory = new FileInfo(appHostDestinationFilePath).Directory.FullName;
-            if (!Directory.Exists(destinationDirectory))
+            byte[] bundleSignature = {
+                // 32 bytes represent the bundle signature: SHA-256 for ".net core bundle"
+                0x8b, 0x12, 0x02, 0xb9, 0x6a, 0x61, 0x20, 0x38,
+                0x72, 0x7b, 0x93, 0x02, 0x14, 0xd7, 0xa0, 0x32,
+                0x13, 0xf5, 0xb9, 0xe6, 0xef, 0xae, 0x33, 0x18,
+                0xee, 0x3b, 0x2d, 0xce, 0x24, 0xb3, 0x6a, 0xae
+            };
+
+            using (var memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostFilePath))
             {
-                Directory.CreateDirectory(destinationDirectory);
-            }
+                using (MemoryMappedViewAccessor accessor = memoryMappedFile.CreateViewAccessor())
+                {
+                    int position = BinaryUtils.SearchInFile(accessor, bundleSignature);
+                    if(position == -1)
+                    {
+                        bundleHeaderOffset = 0;
+                        return false;
+                    }
 
-            // Copy apphost to destination path so it inherits the same attributes/permissions.
-            File.Copy(appHostSourceFilePath, appHostDestinationFilePath, overwrite: true);
+                    bundleHeaderOffset = accessor.ReadInt64(position - sizeof(Int64));
```

Comment:
Perhaps this method should if `bundleHeaderOffset` is zero, like `marker_t::is_bundle()` does.

Reply:
@AustinWise, change is in this commit: https://github.com/dotnet/core-setup/pull/6885/commits/75f19262c4d87f3ed75e1b08e2e7c26fd304cd23

Reply:
Thanks, that should make for a more clear error message.